File: CodingTest/Cpp/Algorithm/practice/quicksort.cpp
#include<iostream>
using namespace std;
void swap(int *a,int *b)
{
    int temp = *a;
    *a =*b;
    *b = temp;
}
void quicksort(int arr[],int left,int right)
{
    if(left>=right) return;
    int pivot = left;
    int start = left+1;
    int end = right;
    while(start<=end)
    {
        while(arr[pivot]>=arr[start] && start<=right) start++;
        while(arr[pivot]<=arr[end] && end>left) end--;
        if(start>end) swap(arr[pivot],arr[end]);
        else swap(arr[start],arr[end]);
    }
    quicksort(arr,left,end-1);
    quicksort(arr,end+1,right);
}
int main()
{
    int arr[] = {0,5,8,4,3,2,6,7,9,1};
    quicksort(arr,0,sizeof(arr)/sizeof(int)-1);
    for(int i=0;i<sizeof(arr)/sizeof(int);i++) cout << arr[i] << " ";
    return 0;
}

File: Language/JavaStudy/study/src/algorithm/QuickSort.java
package algorithm;

import java.util.*;

public class QuickSort {
	private static void swap(int[] arr, int i, int j) {
		int temp = arr[i];
		arr[i] = arr[j];
		arr[j] = temp;
	}

	private static int partition(int[] arr, int low, int high) {
		int pivot = arr[low];
		int i = low;
		int j = high;

		while (i < j) {
			while (i <= high && arr[i] <= pivot) {
				i++;
			}
			while (j >= low && arr[j] > pivot) {
				j--;
			}
			if (i < j) swap(arr, i, j);
		}
		swap(arr, low, j);
		System.out.println(Arrays.toString(arr));
		System.out.println(low + " ~ " + high + ", i: " + i + ", j: " + j);
		return j;
	}

	public static void quickSort(int[] arr, int low, int high) {
		if (low < high) {
			int pivotIndex = partition(arr, low, high);
			quickSort(arr, low, pivotIndex - 1);
			quickSort(arr, pivotIndex + 1, high);
		}
	}

	public static void main(String[] args) {
		int[] arr = {6, 1, 4, 5, 3, 2};
		quickSort(arr, 0, arr.length - 1);
	}
}
